File: backend/app/models/schemas/document.py
"""Pydantic schemas for Document Repository."""

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, ConfigDict

# --- Enums ---


class EntityType(str, Enum):
    """Entity types that documents can be linked to."""

    WBE = "wbe"
    COST_ELEMENT = "cost_element"
    CHANGE_ORDER = "change_order"
    PROJECT = "project"


# --- Folder schemas ---


class DocumentFolderCreate(BaseModel):
    """Properties required for creating a folder."""

    name: str
    parent_id: UUID | None = None


class DocumentFolderUpdate(BaseModel):
    """Properties that can be updated on a folder."""

    name: str | None = None
    parent_id: UUID | None = None


class DocumentFolderPublic(BaseModel):
    """Folder representation returned to the client."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    project_id: UUID
    parent_id: UUID | None
    name: str
    path: str
    created_by: UUID
    created_at: datetime
    updated_at: datetime


# --- Version schemas ---


class DocumentVersionPublic(BaseModel):
    """Document version representation returned to the client."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    document_id: UUID
    version_number: int
    content_type: str
    size_bytes: int
    checksum_sha256: str
    uploaded_by: UUID
    created_at: datetime


# --- Document schemas ---


class DocumentUpdate(BaseModel):
    """Properties that can be updated on a document."""

    name: str | None = None
    description: str | None = None
    tags: list[str] | None = None


class DocumentPublic(BaseModel):
    """Document representation returned to the client."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    project_id: UUID
    folder_id: UUID | None
    name: str
    extension: str
    description: str | None
    tags: list[str]
    current_version: DocumentVersionPublic | None = None
    is_locked: bool
    locked_by: UUID | None
    created_by: UUID
    size_bytes: int
    created_at: datetime
    updated_at: datetime


# --- Entity link schemas ---


class DocumentLinkCreate(BaseModel):
    """Properties required for linking a document to an entity."""

    entity_type: EntityType
    entity_id: UUID
    note: str | None = None


class DocumentLinkUpdate(BaseModel):
    """Properties that can be updated on a document link."""

    note: str


class DocumentLinkPublic(BaseModel):
    """Document entity link representation returned to the client."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    document_id: UUID
    entity_type: str
    entity_id: UUID
    note: str | None
    created_at: datetime


# --- Search ---


class DocumentSearchQuery(BaseModel):
    """Parameters for searching documents."""

    query: str
    folder_id: UUID | None = None
    tags: list[str] | None = None
    extension: str | None = None
    skip: int = 0
    limit: int = 50


# --- Storage stats ---


class StorageStatsPublic(BaseModel):
    """Storage usage statistics for a project."""

    total_bytes: int
    file_count: int
    version_count: int

File: frontend/src/features/documents/types/document.ts
/**
 * Document repository types.
 *
 * These types mirror the backend Pydantic schemas in
 * backend/app/models/schemas/document.py.
 * Once the OpenAPI client is regenerated, these can be replaced by generated types.
 */

// --- Enums ---

export type EntityType = "wbe" | "cost_element" | "change_order" | "project";

// --- Folder ---

export interface DocumentFolderCreate {
  name: string;
  parent_id?: string | null;
}

export interface DocumentFolderUpdate {
  name?: string | null;
  parent_id?: string | null;
}

export interface DocumentFolderPublic {
  id: string;
  project_id: string;
  parent_id: string | null;
  name: string;
  path: string;
  created_by: string;
  created_at: string;
  updated_at: string;
}

// --- Version ---

export interface DocumentVersionPublic {
  id: string;
  document_id: string;
  version_number: number;
  content_type: string;
  size_bytes: number;
  checksum_sha256: string;
  uploaded_by: string;
  created_at: string;
}

// --- Document ---

export interface DocumentUpdate {
  name?: string | null;
  description?: string | null;
  tags?: string[] | null;
}

export interface DocumentPublic {
  id: string;
  project_id: string;
  folder_id: string | null;
  name: string;
  extension: string;
  description: string | null;
  tags: string[];
  current_version: DocumentVersionPublic | null;
  is_locked: boolean;
  locked_by: string | null;
  created_by: string;
  size_bytes: number;
  created_at: string;
  updated_at: string;
}

// --- Entity Link ---

export interface DocumentLinkCreate {
  entity_type: EntityType;
  entity_id: string;
  note?: string | null;
}

export interface DocumentLinkUpdate {
  note: string;
}

export interface DocumentLinkPublic {
  id: string;
  document_id: string;
  entity_type: string;
  entity_id: string;
  note: string | null;
  created_at: string;
}

// --- Storage Stats ---

export interface StorageStatsPublic {
  total_bytes: number;
  file_count: number;
  version_count: number;
}

// --- Outlet context for cost element pages ---

export interface CostElementOutletContext {
  costElement: {
    cost_element_id: string;
    wbe_id: string;
    code: string;
    name: string;
    branch: string;
    [key: string]: unknown;
  } | null;
}
